src/instructlab/eval/common.py

```diff
@@ -0,0 +1,276 @@
+"""
+Common data structures and utilities.
+"""
+
+# Standard
+from typing import Optional
+import ast
+import dataclasses
+import glob
+import json
+import os
+import re
+import time
+
+# Third Party
+from fastchat.model.model_adapter import get_conversation_template  # type: ignore
+import openai
+
+# API setting constants
+API_MAX_RETRY = 16
+API_RETRY_SLEEP = 10
+API_ERROR_OUTPUT = "$ERROR$"
+
+# Categories that need reference answers
+NEED_REF_CATS = ["math", "reasoning", "coding", "arena-hard-200", "taxonomy"]
+
+# Extract scores from judgments
+one_score_pattern = re.compile(r"\[\[(\d+\.?\d*)\]\]")
+one_score_pattern_backup = re.compile(r"\[(\d+\.?\d*)\]")
+
+# Sampling temperature configs for
+temperature_config = {
+    "writing": 0.7,
+    "roleplay": 0.7,
+    "extraction": 0.0,
+    "math": 0.0,
+    "coding": 0.0,
+    "reasoning": 0.0,
+    "stem": 0.1,
+    "humanities": 0.1,
+    "arena-hard-200": 0.0,
+}
+
+
+@dataclasses.dataclass
+class Judge:
+    model_name: str
+    prompt_template: dict
+    ref_based: bool = False
+    multi_turn: bool = False
+
+
+@dataclasses.dataclass
+class MatchSingle:
+    question: dict
+    model: str
+    answer: dict
+    judge: Judge
+    ref_answer: dict = None  # type: ignore[assignment]
+    multi_turn: bool = False
+
+
+def bench_dir(output_dir, bench_name, branch) -> str:
+    b_dir = f"{output_dir}/{bench_name}"
+    if branch is not None:
+        b_dir = os.path.join(b_dir, branch)
+    return b_dir
+
+
+def load_questions(question_file: str, begin: Optional[int], end: Optional[int]):
+    """Load questions from a file."""
+    questions = []
+    with open(question_file, "r", encoding="utf-8") as ques_file:
+        for line in ques_file:
+            if line:
+                questions.append(json.loads(line))
+    questions = questions[begin:end]
+    return questions
+
+
+def load_model_answers(answer_dir: str, model_name=None):
+    """Load model answers.
+
+    The return value is a python dict of type:
+    Dict[model_name: str -> Dict[question_id: int -> answer: dict]]
+    """
```

Comment:
Ret value can also be specified in type hint, e.g.

```python
def load_model_answers(answer_dir: str, model_name=None) -> dict:

Reply:
Done